Libraries/src/net/relinc/libraries/data/ModifierFolder/Resampler.java

```diff
@@ -0,0 +1,163 @@
+		return "Data Resampler";
+	}
+
+	@Override
+	public ModifierResult applyModifier(double[] x, double[] y, DataSubset data) {
+		if(shouldApply()) {
+			double[] newX = this.sampleData(x, this.userDataPoints);
+			double[] newY = this.applyModifierToData(x, y, data);
+			return new ModifierResult(newX, newY, this.getUserIndexToOriginalIndexRatio(x.length));
+		}
+		return new ModifierResult(x, y, 1.0);
+	}
+
+	private double getUserIndexToOriginalIndexRatio(int originalLength) {
+		return (this.userDataPoints - 1.0) / (originalLength - 1.0);
+	}
+
+	public double[] applyModifierToData(double[] time, double[] fullData, DataSubset activatedData) {
+		// If the fitter polynomial modifier is activated, we should use that for interpolation.
+		// Else, do linear interpolation between the two closest points.
+
+
+		Fitter fitter = activatedData == null ? null : activatedData.getModifiers().getFitterModifier(); // kinda weird
+		if(fitter != null && fitter.activated.get() && fitter.enabled.get()) {
+
+//			FitableDataset fitable = new FitableDataset();
+//			fitable.setBeginFit((int)(fitter.fitable.getBeginFit() * this.getUserIndexToOriginalIndexRatio(time.length)));
+//			fitable.setEndFit((int)(fitter.fitable.getEndFit() * this.getUserIndexToOriginalIndexRatio(time.length)));
+//			fitter.applyModifierToData(time, fullData, fitable); // this has side effects on `fitter`
+
+
+			double[] sampledTime = sampleData(time, userDataPoints);
+			fullData = sampleData(fullData, userDataPoints);
+			int start = (int)(fitter.fitable.getBeginFit() * this.getUserIndexToOriginalIndexRatio(time.length));
+			int end = (int)(fitter.fitable.getEndFit() * this.getUserIndexToOriginalIndexRatio(time.length));
+			for(int i = start; i <= end; i++) {
+				fullData[i] = fitter.fitable.computeY(sampledTime[i]);
+			}
+		} else {
+			fullData = sampleData(fullData, userDataPoints);
+		}
+
+		return fullData;
+
+	}
+
+	@Override
+	public List<Node> getTrimDataHBoxControls() {
+		ArrayList<Node> list = new ArrayList<>();
+		list.add(holdGrid);
+		return list;
+	}
+
+	@Override
+	public String getStringForFileWriting() {
+		return enabled.get() ? resamplerDescription + ":" + userDataPoints + SPSettings.lineSeperator : "";
+	}
+
+	@Override
+	public void setValuesFromDescriptorValue(String descrip, String val) {
+		if(descrip.equals(resamplerDescription)){
+			//it was saved, so it is enabled
+			userDataPoints = Integer.parseInt(val);
+			enabled.set(true);
+			activated.set(true);
+		}
+	}
+
+	@Override
+	public void readModifierFromString(String line) {
+		setValuesFromLine(line);
+	}
+
+	@Override
+	public void configureModifier(DataSubset dataSubset) {
+		userDataPoints = valueTF.getDouble().intValue();
+	}
+
+	public double getUserDataPoints() {
+		return userDataPoints;
+	}
+
+	public void setUserDataPoints(int userDataPoints) {
+		this.userDataPoints = userDataPoints;
+	}
+
+
+	public static double[] sampleData(double[] source, int numDataPoints) {
```

Comment:
@bhalonen A good thing for you to look at. This does linear interpolation between the 2 closest points

Reply:
The code that calls this function uses the fitter modifier if appropriate